.travis.yml

```diff
@@ -1,102 +1,173 @@
+    # that jobs, because "artifacts" need to be rebuilt which takes more than
+    # 50 min and thus hits the time limit for non-private jobs and gets killed).
+    - SMART_CACHE=1
+    # Prevent building docs. Uncomment to build them.
+    - NO_DOCS=1
+    # Upload a copy of the build dir before and after syncing with the repo
+    #- MKDBG=1
+
+    # Uncomment the next line to build w/o stack
+    - STACK_ROOT=/var/tmp/stack
+
+    # Here we build and cache builds
+    - HETS_BASEDIR=/var/tmp/hets HETS_BUILD_DIR=${HETS_BASEDIR}/build
+
+    - PREFIX=/tmp/hets-install
+    - PATH=${PREFIX}/bin:${STACK_ROOT}/bin:${PATH}
+    - HETS_LIB=/var/tmp/Hets-lib
+    - HETS_MAGIC=${TRAVIS_BUILD_DIR}/magic/hets.magic
+    - TH=utils/travis-helper.sh
+
+
+cache:
+  # we push a lot, so 180s are sometimes not enough ...
+  timeout: 600
+  directories:
+  - ${STACK_ROOT:-/empty}
+  - ${PKG_CACHE:-/empty}
+  - ${HETS_LIB}
+  - ${HETS_BASEDIR}
+
 install:
-  # Download and unpack the stack executable
-  - mkdir -p ~/.local/bin
-  - export PATH=$HOME/.local/bin:$PATH
-  - travis_retry curl -L https://www.stackage.org/stack/linux-x86_64 | tar xz --wildcards --strip-components=1 -C ~/.local/bin '*/stack'
-  - sudo apt-get install --no-install-recommends `dpkg-checkbuilddeps debian/control 2>&1 | cut -f3- -d":" | sed -e 's,([^)]*),,g' -e 's,openjdk-.*-jdk,,'` || true
-
-before_script:
-  - export HETS_LIB=$HOME/Hets-lib
-  - export HETS_MAGIC=$PWD/magic/hets.magic
-  - |
-    if [ -d $HETS_LIB/.git ]
-    then
-      pushd $HETS_LIB
-      git pull
-      popd
-    else
-      git clone --depth=1 https://github.com/spechub/Hets-lib.git $HETS_LIB
-    fi
-
-  # For some reason, Travis CI sets this environment variable, which breaks the
-  # tests (_JAVA_OPTIONS=-Xmx2048m -Xms512m)
-  - unset _JAVA_OPTIONS
-
-  # Compile Hets
-  - make stack
-  - stack exec -- ghc -V
-
-matrix:
+  - set >/tmp/set
+  - date '+%s' >/tmp/timestamp ; date
+  - ${TH} myShowCacheUsage
+  # because the context is important, we need to call 'set' explicitly _here_
+  - ls -al ${HOME}/.casher
+  - ${TH} myShowInfo short
+  - ${TH} myCheckStack
+  - ${TH} myGetMissingPkgs
+  - ${TH} myCheckHetslib
+
+before_cache:
+  - ksh -c 'printf "%(%s)T\n" now ; printf "%T\n" now'
+  - ${TH} myBefore_cache
+  - ${TH} myShowCacheUsage
+
+after_script:
+  - ksh -c 'printf "%(%s)T\n" now ; printf "%T\n" now'
+  - ls -al ${HOME}/.casher
+
+
+jobs:
   include:
-    - env: JOB='make check'
+    # .1
+    - stage: Stack
+      if: env(STACK_ROOT) IS present AND env(SKIP_STACK) != 1
+      before_install:
+        - echo stack >/tmp/JOB_NAME
       script:
-        - make
-        - ./hets -V
-        # Run the test
+        - ${TH} myPrepareBuildDir
+        - cd ${HETS_BUILD_DIR}
+        # Compile Hets
+        - make stack && touch ${STACK_ROOT}/bin/stack.done
+        - stack exec -- ghc -V
+
+
+    # .2
+    - stage: Build
+      if: type != push AND env(SKIP_BUILD) != 1
```

Comment:
Why are the build and test stages not run when code is pushed?

Reply:
When the PR gets pushed (e.g. to the master), builds and tests have been run already, so no need to run/delay it again for 50+ min. Also there is usually no value in caching build results, because PRs should always do a complete, clean build to avoid usage of old artifacts from previous builds.

However, if this works as I expect wrt. the shallow travis-ci docs, PRs will benefit from building Stack on push, because all PRs can use the same cached Stack build. If one thinks, stack should be rebuild, one can manually delete the caches via travis-ci web ui, and all starts from zero.

Reply:
Ok this makes sense 👍.